components/brave_today/browser/brave_news_p3a.cc

```diff
@@ -115,13 +116,44 @@ void RecordWeeklyAddedDirectFeedsCount(PrefService* prefs, int change) {
                           weekly_total);
 }
 
+void ResetCurrSessionTotalViewsCount(PrefService* prefs) {
+  prefs->SetUint64(prefs::kBraveTodayCurrSessionCardViews, 0);
+  VLOG(1) << "NewsP3A: reset curr session total card views count";
+}
+
+void RecordTotalCardViews(PrefService* prefs,
+                          uint64_t cards_viewed_session_total_count) {
+  WeeklyStorage total_storage(prefs, prefs::kBraveTodayTotalCardViews);
+
+  uint64_t stored_curr_session_views =
+      prefs->GetUint64(prefs::kBraveTodayCurrSessionCardViews);
+
+  // Since the front-end repeatedly sends the updated total,
+  // we should subtract the last known total for the current session and
+  // add the new total.
+  total_storage.SubDelta(stored_curr_session_views);
```

Comment:
Saturating at 0 here is equivalent to the old update being removed having aged out of the WeeklyStorage retention window, so we're not losing anything?

Reply:
The stored "current session views" preference is cleared upon the start of a new session (see line 149). Values recorded on previous days will not be lost as a result.

Reply:
There is a problem with the architecture of storing data about a "current" Brave News session as a Profile Pref, and then in this function expecting the incoming data to be part of the same session stored in the pref. Brave News can have multiple active sessions, and the user can go between them. For example, they can return to a session and keep scrolling. So I think the code in this PR has introduced bad data to P3A where it may be subtracting the view count from 1 session and adding the view count from another.

This is why Brave News' frontend carefully makes sure to preserve local state in the UI for each UI instance, even when navigating forward to an article and then back to a Brave News feed. It sends that state to the backend when reporting metrics. If we want to store anything additional in state, or send incremental changes instead, that should change should be made within each session instance.